### ISSUE TITLE
[HOLD for payment 2022-12-20] Web- Chat - Onyx error is displayed in console when requesting money

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#12432]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to staging.new.expensify.com
2. Logged in with applausetester+0901abb@applause.expensifail.com/Feya86Katya
3. Open Console
4. Request money from kbecciv+1103abb1@gmail.com/Feya86Katya

## Expected Result:
Onyx error is not displayed in console when requesting money

## Actual Result:
Onyx error is displayed in console when requesting money

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Web

**Version Number:** 1.2.23.7

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**Email or phone of affected tester (no customers):** 

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation
![image (26)](https://user-images.githubusercontent.com/93399543/199779085-af45843d-c4ed-4bb5-8c4d-eeedb671d58d.png)

https://user-images.githubusercontent.com/93399543/199779111-b7cb9785-926b-4d70-8477-0795e75933d9.mp4

**Expensify/Expensify Issue URL:** 

**Issue reported by:** Applause - Internal Team

**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by johncschuster** [comment]
I couldn't reproduce the Onyx error when I followed the repro steps.

Here's what I did:
Go to staging.new.expensify.com
Logged in with [applausetester+0901abb@applause.expensifail.com](https://www.expensify.com/_support/?supportEmail=applausetester%2B0901abb%40applause.expensifail.com&comment=better%20supportal%20issue%20nubmer%3A%2012432&location=_support/)/Feya86Katya
Open Console
Request money from [kbecciv+1103abb1@gmail.com](https://www.expensify.com/_support/?supportEmail=kbecciv%2B1103abb1%40gmail.com&comment=better%20supportal%20issue%20nubmer%3A%2012432&location=_support/)/Feya86Katya

And here's the result:

![2022-11-03_16-14-34](https://user-images.githubusercontent.com/5579997/199835689-29d48cfb-103e-4c0f-95a4-35ac1856a448.png)

**Comment by johncschuster** [comment]
@kbecciv Can you try reproducing this again? If you continue to get this behavior, let's see if we can clarify the repro steps.

**Comment by kbecciv** [comment]
@johncschuster Checking again with latest build, will update you shortly.

**Comment by mallenexpensify** [comment]
Bumped back to daily @johncschuster , per the new BZ process.

**Comment by johncschuster** [comment]
@amyevans, I'm trying to decide if this can be labeled `Internal` or `External`. Do you have a sense of which direction this should go?

**Comment by amyevans** [comment]
I haven't attempted to reproduce myself yet, but given that it's related to data from the API I think it's best to keep it `Internal`. Could you try reproducing again @johncschuster and writing up more specific repro steps if successful? I can also take a closer look Wednesday, if not before (but it's getting late for me today and I'm traveling tomorrow, so I don't want to overpromise).

**Comment by johncschuster** [comment]
Thanks for the buddy check, Amy! I'll label it `Internal` and will try reproducing the behavior again.

**Comment by johncschuster** [comment]
I wasn't able to get back to testing this today. I'll carve out some time tomorrow morning.

**Comment by amyevans** [comment]
I spent some time with this today and can reliably reproduce locally with the following steps:

1. Create two IOU requests
2. Cancel both of them
3. Create another IOU request
4. Observe error in console

To fix the Onyx error (`An error occurred while applying merge for key: reportIOUs_XYZ Error: TypeError: Cannot convert undefined or null to object`), we need to call the Onyx `set` method instead of `merge` in [this line](https://github.com/Expensify/Web-Expensify/blob/cf6e14ae37542719cee69641b9fdd1b3b3c2954a/lib/TransactionAPI.php#L936). I've opened a PR for that: https://github.com/Expensify/Web-Expensify/pull/35526.

However I'm unsure what to do, if anything, about the second error (`Report no longer exists`). I see it occurs because we call `Report.fetchIOUReportByID()` for the optimistic `iouReportID` which doesn't exist on the server, hence it's expected to be not found (since we're using the other `iouReportID` that already exists instead). But beyond that level of understanding, I'm getting a bit lost in the code. Hoping maybe @Gonals you can offer an opinion/direction, since it looks like you worked on this refactor originally. Thanks! 🙏

**Comment by Gonals** [comment]
I haven't dug a lot into the code, but I think the key here is that there's only ever one active IOU report in a chat.
What I believe is happening is that, when we delete the requests, the IOUReport gets deleted in the backend, but the frontend is not informed and keeps trying to use it. This should be the cause of both issues, so I don't think modifying [this line](https://github.com/Expensify/Web-Expensify/blob/cf6e14ae37542719cee69641b9fdd1b3b3c2954a/lib/TransactionAPI.php#L936) is the way to go to fix the underlying problem.

Basically, I believe we are trying to add a Money Request to an IOU report that no longer exists. Chances are we are missing an onyx update the moment we delete the last request

**Comment by johncschuster** [comment]
Thanks for digging into this @amyevans and @Gonals! In light of what you've found, do you think this is a proper bug in need of fixing?

**Comment by Gonals** [comment]
I believe so, yes :)

**Comment by amyevans** [comment]
Yeah, definitely a bug! I'm continuing to dig into it (just didn't have much time on Friday after PR reviews/other chores).

**Comment by johncschuster** [comment]
Sweet! Thanks for taking it on, @amyevans! I'm going to go ahead and unassign myself. Let me know if you need me to jump back in!

**Comment by amyevans** [comment]
After digging in deeper, I see the current flow for a request > cancel > request is loosely:
- No `reportIOUs_123` Onyx key to start
- `requestMoney()` creates an optimistic IOU report ID on the FE https://github.com/Expensify/App/blob/64eb889dc42ecab504e62d354aca4f7541511ef3/src/libs/actions/IOU.js#L75
- The server pushes Onyx updates including a merge into the `reportIOUs_123` key, another is a merge into the report key with
```js
{
  hasOutstandingIOU: true,
  iouReportID: "123"
}
```
- `cancelMoneyRequest()` -  it correctly finds the existing `iouReportID` and updates the data
-  The server pushes Onyx updates like before, including:
```js
{
  hasOutstandingIOU: false,
  iouReportID: null
}
```
- On subsequent `requestMoney()`, the FE creates a _different_ optimistic reportIOU instead of reusing the existing, due to this code: https://github.com/Expensify/App/blob/64eb889dc42ecab504e62d354aca4f7541511ef3/src/libs/actions/IOU.js#L67-L76

So I think to fix the bug we can:
* Update the API to return the existing `iouReportID` even when `hasOutstandingIOU` is `false` [here](https://github.com/Expensify/Web-Expensify/blob/b338229b2c9606968324a70b6b52209685c0b1ad/lib/ReportUtils.php#L1814), plus
* Update [this condition](https://github.com/Expensify/App/blob/64eb889dc42ecab504e62d354aca4f7541511ef3/src/libs/actions/IOU.js#L68) to check for `if (chatReport.iouReportID)`

This works in local testing and makes sense to me that once we've got an `iouReportID` we should continue to reuse it, but hoping you can gut check this solution @Gonals before I officially put up the 2 PRs. Thanks!

**Comment by amyevans** [comment]
I have draft PRs up, but in testing beyond this exact use case, I've uncovered the same bug in a slightly different flow:
1. Request money from the big `+` button and choose someone with NO chat with you.
2. Cancel the money request.
3. Create another money request with that user.

My changes do not fix that issue, so it must be a slightly different root cause. I'm still investigating what exactly is happening there and how to fix it.

**Comment by amyevans** [comment]
Still plugging away at this, so far I'm able to see that the optimistic report ID and IOU ID are correctly created on the first money request, and the data looks fine through the cancel request. But somewhere between that point and the next money request, the `iouReportID` property on the report in Onyx is reset to 0. Still trying to trace if that's coming from another API request and subsequent Onyx push where the report object is incomplete, or if there's something happening in the FE with stale data or something? Feels like the solution is within reach but I need to put this down again for today, will pick it back up Monday if not sooner.

**Comment by amyevans** [comment]
Not overdue, working on it as my top priority today

**Comment by amyevans** [comment]
In further investigation, it looks like after the initial money request and cancellation pair, `OpenReport` is called and is the API command that's causing the `iouReportID` property on the report in Onyx is reset to `0`. `OpenReport` calls `Auth::getChats()` [here](https://github.com/Expensify/Web-Expensify/blob/2c80495239e924d7be19f01194b4777f78d6bbac/lib/ReportAPI.php#L4061), which in Auth [here](https://github.com/Expensify/Auth/blob/main/auth/command/GetChats.cpp#L157-L162) appears to be very deliberately only returning an `iouReportID` if the outstanding total is non zero.

So based on that information, it feels like my initial proposed solution is wrong. We shouldn't reuse the same `iouReportID` over and over, instead we should generate a new one if the outstanding balance is zero (the last IOU Report is no longer "active"). Does that sound right? (If so, that kind of leaves me back at square one trying to understand what to do here...)

Bump @Gonals for some feedback please 🙏 , or cc @mountiny @Julesssss since l think you're familiar with this area as well.

**Comment by mountiny** [comment]
Gonna try to look tomorrow.

**Comment by Gonals** [comment]
> instead we should generate a new one if the outstanding balance is zero

I believe this may be correct. At least, we seem to generate a new one with the previous one is "done"

**Comment by mountiny** [comment]
> We shouldn't reuse the same iouReportID over and over, instead we should generate a new one if the outstanding balance is zero (the last IOU Report is no longer "active"). Does that sound right? (If so, that kind of leaves me back at square one trying to understand what to do here...)

@amyevans That is correct, there should be always one active IOU report connected to the chat report and I assume cancelling everything on the report so there is nothing outstanding can have same effect as settling the report.

We are not returning the IOU report ids with total 0 on purpose. cc @Julesssss 

Not sure if I followed exactly the entire thread here, but seems like we need to ensure to create optimistic IOU report when there are some non-settled but with 0 total.

**Comment by amyevans** [comment]
Ok thanks @Gonals and @mountiny ❤️ 

> seems like we need to ensure to create optimistic IOU report when there are some non-settled but with 0 total.

The optimistic IOU report does get created on the FE. In Web-E we then call `Auth::createIOUTransaction()` with the new `iouReportID`, but Auth returns the original IOU report ID:

![response-from-auth](https://user-images.githubusercontent.com/7277067/203370265-1f57ce5b-e9f9-4288-a050-91a067c87a27.png)

Trying to dig into Auth at the moment to understand why that is.

**Comment by mountiny** [comment]
Uh ok, maybe we should make sure we are all laigned on what should happen with these IOU reports this were not settled but cancelled to total of 0. Maybe we wnat ot reuse it and in that case we have to send them back to the frontend. 

Would you be able to start a thread in Slack, I think @Julesssss and maybe Carlos Sr can help with this the best.

**Comment by amyevans** [comment]
We discussed [in Slack](https://expensify.slack.com/archives/C03TQ48KC/p1669138744156809) and determined that the existing `iouReportID` should be reused. To prevent the issue where the existing `iouReportID` is overwritten on `OpenReport`, we'll only include that key in the response if there is a non-zero ID returned from Auth.

I updated the 2 PRs with this and tested pretty extensively locally. In merging `main` to my branches, I discovered @Gonals has some existing PRs that change the Onyx key where we store IOU data ([Web-E PR](https://github.com/Expensify/Web-Expensify/pull/35594), [App PR](https://github.com/Expensify/App/pull/12780)), so if testing the existing combo of my Web-E branch + App branch, there are some console errors related to the mismatch in keys. But that'll clear up once https://github.com/Expensify/App/pull/12780 is deployed.

**Comment by Gonals** [comment]
> We discussed [in Slack](https://expensify.slack.com/archives/C03TQ48KC/p1669138744156809) and determined that the existing `iouReportID` should be reused. To prevent the issue where the existing `iouReportID` is overwritten on `OpenReport`, we'll only include that key in the response if there is a non-zero ID returned from Auth.
> 
> I updated the 2 PRs with this and tested pretty extensively locally. In merging `main` to my branches, I discovered @Gonals has some existing PRs that change the Onyx key where we store IOU data ([Web-E PR](https://github.com/Expensify/Web-Expensify/pull/35594), [App PR](https://github.com/Expensify/App/pull/12780)), so if testing the existing combo of my Web-E branch + App branch, there are some console errors related to the mismatch in keys. But that'll clear up once #12780 is deployed.

Nice! Which keys are you using in your PRs? Are they on HOLD because of mine?

**Comment by luacmartins** [comment]
Related to this issue, now that we are storing IOU reports in the `reports_` key, we are actually trying to navigate to an IOU report as if they were chat reports, which seems to be (one?) of the sources for this console error and this deploy blockerhttps://github.com/Expensify/App/issues/13000. I put up a fix [here](https://github.com/Expensify/App/pull/13028) to filter out IOU reports from the initial loading screen.

**Comment by kbecciv** [comment]
@luacmartins Can you please confirm if its the same root of cause, when clicking on Settled/Request options? or we need to create a separate GH link for it?

https://user-images.githubusercontent.com/93399543/204012336-f43e30fe-a678-48d1-8bd3-5c398d712ca0.mp4

**Comment by luacmartins** [comment]
@kbecciv that seems like a different issue to me. I logged the [issue](https://github.com/Expensify/App/issues/13048) and put up a fix for it [here](https://github.com/Expensify/Web-Expensify/pull/35674)

**Comment by amyevans** [comment]
Not overdue, the 2 PRs were just updated and are under review.

**Comment by mountiny** [comment]
Assigned @mananjadhav and @sakluger to manage a payment for this PR which Manan has reviewed and tested. that will be once the PR hits production

**Comment by sakluger** [comment]
@mountiny @amyevans can you catch me up on who needs to be paid here and how much? I see it's labeled as Internal, so just making sure I handle this properly since I've only ever paid out reporting bonuses on internal issues before.

**Comment by mountiny** [comment]
@sakluger of course, at the moment we are paying flat rate of $1000 to C+ who review and test our internal PRs. For this issue, we should pay @mananjadhav $1000 for his review and testing 🙇 

We just need to make a job for $1000 for Manan, not sure what we usually write into these job descriptions.

**Comment by sakluger** [comment]
@mananjadhav I've created the Upwork post and sent you an offer.

Here's the Upwork post: https://www.upwork.com/ab/applicants/1604898467295248384/job-details (external: https://www.upwork.com/jobs/~011f05def8834dffeb).

**Comment by mananjadhav** [comment]
Thanks for the invite, accepted the offer

**Comment by sakluger** [comment]
@amyevans @Gonals do you think this needs a new regression test? And if so, would we add steps to:

- [Global Create - Split Bill - IOU Preview](https://expensify.testrail.io/index.php?/cases/view/1971030&group_by=cases:section_id&group_order=asc&display_deleted_cases=0&group_id=229063) test case
- [IOU - Cancelling/Declining Requests](https://expensify.testrail.io/index.php?/cases/view/1971037&group_by=cases:section_id&group_order=asc&display_deleted_cases=0&group_id=229063) test case

From reading through the issue, it is an error in the IOU creation, but I think it only happens after you'd deleted a previous request?

**Comment by sakluger** [comment]
Paid @mananjadhav in Upwork! Just waiting to see if we need to add any new regression tests.

**Comment by sakluger** [comment]
No regression tests needed since this is just a console error.
